### ISSUE TITLE
Patrones: Tarea 03

### ISSUE THREAD

**Issue by domingomery** [domingomery/patrones#22]
Por favor hacer comentarios aquí. Respondo todas las preguntas siempre y cuando no las haya aclarado en clases, o no estén escritas en el enunciado de la Tarea, o no las haya contestado en este foro. Usemos el foro responsablemente para que sea de utilidad. Preguntas repetidas, pueden hacer este foro excesivamente largo, lo que dificultará su lectura. Muchas gracias! Domingo

**Comment by cristianhdez** [comment]
Profesor, se puede usar keras en vez de sklearn?

**Comment by domingomery** [comment]
> Profesor, se puede usar keras en vez de sklearn?

si, se puede, pero nada de deep learning (nada de ReLu, dropout, MaxPool, etc.), y no más de dos capas ocultas...

**Comment by VicenteMerino** [comment]
Hola Profe, en este [colab](https://colab.research.google.com/drive/1_xE4tTQwtyiE_9zjLkEssJMYTGytgZZg?usp=sharing) en la 5a celda de código, me sale que `nn_st` no está definido. Qué debería tomar ese parámetro?

**Comment by domingomery** [comment]
> Hola Profe, en este [colab](https://colab.research.google.com/drive/1_xE4tTQwtyiE_9zjLkEssJMYTGytgZZg?usp=sharing) en la 5a celda de código, me sale que `nn_st` no está definido. Qué debería tomar ese parámetro?

ups! es un string para que aparezca en la grafica, puedes poner nn_st = 'LR' o bien:

`show_clf_results(clf,Xtrain,ytrain,Xtest,ytest,y0,ypred,'LR',decisionline=1)`

**Comment by VicenteMerino** [comment]
Profesor, una consulta, me di cuenta que la base de datos actual, a diferencia de la mayoría que vimos para la extracción de características, no está en escala de grises, por lo que al hacer plt.imread, me arroja una matriz 3D (uno por cada canal RGB), lo cual es incorrecto si quiero usar las librerías de pybalu (pues requiere matrices 2D). Por lo anterior me surgen un par de dudas:

- Es válido pasar la imagen en escalas de grises? Si es así, vi este código en el ejemplo de las vacas con SIFT, donde se usa 
```python
I = cv2.cvtColor(cv2.imread(st, cv2.COLOR_BGR2RGB)
J= cv2.cvtColor(I, cv2.COLOR_RGB2GRAY)
``` 
Esto es correcto para pasar a grises?
- Qué otra alternativa habría? Me da miedo que al pasar a escala de grises se pueda perder mucha información.

**Comment by domingomery** [comment]
> Es válido pasar la imagen en escalas de grises? 

Sí, es válido.

> Esto es correcto para pasar a grises?

Tienes que probarlo, se ve bien en todo caso.

> 
> Qué otra alternativa habría? 

Puedes leer adicionalmente cada canal de la imagen a color y tratarlo como si fuera una imagen nueva e independiente. De esta manera extraes features y concatenas. Como ejemplo si sacas LBP de 59 elementos (y sin particiones), tendrías a tu disposición 59 features del canal Rojo, 59 del canal Verde, 59 del canal Azul y 59 de la conversión a grises, es decir 59 x 4 = 236 features que podrías luego seleccionar. Intuyo que el canal Rojo tiene información relevante.

**Comment by fnberrios** [comment]
Hola profesor, estuve haciendo un listado de todas las imágenes del training de la base de datos y al parecer la clase 0 posee 1351 imágenes en lugar de 1350. ¿Hay algún inconveniente con esto?

**Comment by domingomery** [comment]
> Hola profesor, estuve haciendo un listado de todas las imágenes del training de la base de datos y al parecer la clase 0 posee 1351 imágenes en lugar de 1350. ¿Hay algún inconveniente con esto?

Qué extraño... he bajado la base de datos del enunciado y veo que hay 1350 imágenes en el training de la clase 0 y también de la clase 1... puedes revisar de nuevo cuantos archivos `jpg` hay? a veces en los directorios el sistema operativo incluye archivos adicionales (que no son `jpg`). O quizás se te duplicó por error una de las imágenes. Me cuentas, en todo caso no es grave.

**Comment by fnberrios** [comment]
> > Hola profesor, estuve haciendo un listado de todas las imágenes del training de la base de datos y al parecer la clase 0 posee 1351 imágenes en lugar de 1350. ¿Hay algún inconveniente con esto?
> 
> Qué extraño... he bajado la base de datos del enunciado y veo que hay 1350 imágenes en el training de la clase 0 y también de la clase 1... puedes revisar de nuevo cuantos archivos `jpg` hay? a veces en los directorios el sistema operativo incluye archivos adicionales (que no son `jpg`). O quizás se te duplicó por error una de las imágenes. Me cuentas, en todo caso no es grave.

Efectivamente fue eso, hay un archivo .DS_Store que estaba considerando en el directorio. Gracias!

**Comment by VicenteMerino** [comment]
> > > Hola profesor, estuve haciendo un listado de todas las imágenes del training de la base de datos y al parecer la clase 0 posee 1351 imágenes en lugar de 1350. ¿Hay algún inconveniente con esto?
> > 
> > 
> > Qué extraño... he bajado la base de datos del enunciado y veo que hay 1350 imágenes en el training de la clase 0 y también de la clase 1... puedes revisar de nuevo cuantos archivos `jpg` hay? a veces en los directorios el sistema operativo incluye archivos adicionales (que no son `jpg`). O quizás se te duplicó por error una de las imágenes. Me cuentas, en todo caso no es grave.
> 
> Efectivamente fue eso, hay un archivo .DS_Store que estaba considerando en el directorio. Gracias!

Repecto a esto mismo está bien si subimos nuestro propio zip sin ese archivo, para hacer más limpia la lectura?

**Comment by domingomery** [comment]
> Repecto a esto mismo está bien si subimos nuestro propio zip sin ese archivo, para hacer más limpia la lectura?

no se puede, deben usar el mismo link, es la única forma de asegurar que todo/as trabajen con exactamente los mismos datos.

**Comment by domingomery** [comment]
> Hola profesor, la carga de la base de datos para la tarea 3 no es tan trivial como antes ya que los nombres de las fotos en las carpetas no son seguidos. ¿Qué podemos hacer? ya que me imagino que la idea de la tarea no es perder tiempo en esto. Las funciones num2fixstr y imageLoad ya no son tan útiles y no entiendo como llamar carpetas dentro de carpetas en el .zip. Gracias!!

He hecho un [colab](https://colab.research.google.com/drive/1yyY7SFDjkWXv0AdP8Bn60fC-S-3VEarO?usp=sharing) donde se puede extraer características de todas las imágenes de un directorio.

**Comment by fnberrios** [comment]
Hola profe! Consulta, ¿es posible usar CNN como clasificador? Lo pregunto porque lo vimos en la clase de Introducción a Deep Learning y se supone que no podemos usar esas técnicas, pero en la misma clase, al mostrar este ejemplo del [https://colab.research.google.com/drive/14I4sEnSZYYGrejdSWpuxFX0CLSv4oJtK?usp=sharing#scrollTo=Yh_ldAFZEPzK](url), mencionó que podríamos hacer algo similar en la tarea. Gracias de antemano.

**Comment by fnberrios** [comment]
Otra duda, ¿hay un límite de nodos que pueden tener las capas ocultas de una red neuronal?

**Comment by domingomery** [comment]
> Hola profe! Consulta, ¿es posible usar CNN como clasificador? 

No se puede usar CNN (ya que esto es deep learning). En la clase de Deep Learning mencioné un ejemplo de Redes Neuronales (NN) en el que el input es una imagen de 28x28 en tonos de gris, que se conecta a varias capas ocultas de 64 nodos y una capa de saluda de 10 nodos (ver ejemplo PAT04_NN_MNIST, ojo que no es el ejemplo PAT04_CNN_MNIST). Es el ejemplo NN para la base de datos MNIST (explicado a partir del min 13 del video de la clase). Para imágenes pequeñas este método basado en NN (no en CNN) puede funcionar, ya que el entrenamiento es posible (no son tantos parámetros). En la tarea está permitido usar una Red Neuronal (no una CNN) de dos capas ocultas en el que el input sea la imagen, pero no lo recomiendo porque las imagenes son muy grandes y con dos capas solamente es dificil obtener resultados aceptables. Si bajas el tamaño de la imagen, probablemente la NN pueda entrenarse pero no creo que se obtengan buenos resultados porque pierdes muchos detalles de la piel.

**Comment by domingomery** [comment]
> Otra duda, ¿hay un límite de nodos que pueden tener las capas ocultas de una red neuronal?

No hay límite

**Comment by fnberrios** [comment]
@domingomery Gracias!

**Comment by cristianhdez** [comment]
Hola profesor,
Para el primer bloque, cuenta como bloques distintos usar las mismas características pero de distintos canales de la imagen (o de la imagen en escala de grises)?

**Comment by domingomery** [comment]
> Para el primer bloque, cuenta como bloques distintos usar las mismas características pero de distintos canales de la imagen (o de la imagen en escala de grises)?

No, no es la idea.

**Comment by rodrigohanuch** [comment]
> > Para el primer bloque, cuenta como bloques distintos usar las mismas características pero de distintos canales de la imagen (o de la imagen en escala de grises)?
> 
> 
> 
> No, no es la idea.
> 
> 
> 
> 

Y cuenta hacer la combinación de los grises CON colores?

**Comment by VicenteMerino** [comment]
Hola profe, es válido hacer cleaning de los datos?

**Comment by VicenteMerino** [comment]
Al tratar de ajustar la cantidad de características que son seleccionadas por sfs, esto también tiene que ser hecho con el conjunto de validación, o en este caso puede ser con el de training?

**Comment by domingomery** [comment]
> Hola profe, es válido hacer cleaning de los datos?

no sé bien a que te refieres... 

- si se puede usar `clean`? claro que se puede usar, y es recomendable

- si quieres hacer procesamiento de imágenes antes de extraer features? también se puede hacer

- si quieres eliminar imágenes de la base de datos de manera manual? no se puede hacer 

- si quieres eliminar algunas imágenes del training de manera automática (es decir si no quieres imágenes en el training que tengan poco contraste por ejemplo, y el contraste es una feature que le mides a la imagen)? si se puede hacer 

- no se puede eliminar ninguna imagen del testing

**Comment by domingomery** [comment]
> Al tratar de ajustar la cantidad de características que son seleccionadas por sfs, esto también tiene que ser hecho con el conjunto de validación, o en este caso puede ser con el de training?

ver enunciado

**Comment by domingomery** [comment]
> > > Para el primer bloque, cuenta como bloques distintos usar las mismas características pero de distintos canales de la imagen (o de la imagen en escala de grises)?
> > 
> > 
> > No, no es la idea.
> 
> Y cuenta hacer la combinación de los grises CON colores?

No entiendo la pregunta, puedes ser más específico?

**Comment by sebapulgar** [comment]
Hola profe, 
Para esta tarea, el tiempo de extracción de características está siendo mucho mayor al de las tareas anteriores llegando a 20 o 25 minutos, ya que estoy extrayendo varios tipos de características para todos los canales de color, es válido que se demore tanto? Pensando en que después hay que revisar cada tarea.
Saludos

**Comment by rodrigohanuch** [comment]
> > > > Para el primer bloque, cuenta como bloques distintos usar las mismas características pero de distintos canales de la imagen (o de la imagen en escala de grises)?
> 
> > > 
> 
> > > 
> 
> > > No, no es la idea.
> 
> > 
> 
> > Y cuenta hacer la combinación de los grises CON colores?
> 
> 
> 
> No entiendo la pregunta, puedes ser más específico?

Por ejemplo usar una combinación de diferentes canales pero con las mismas características, digamos en el primer caso uso gris, y en segundo Rojo y el canal Cr o Y de YCrCb.

Al final se usan características diferentes de la imagen porque el canal Cr o Y representan otras cosas que el Grey scale por si solo

**Comment by domingomery** [comment]
> Por ejemplo usar una combinación de diferentes canales pero con las mismas características, digamos en el primer caso uso gris, y en segundo Rojo y el canal Cr de YCrCb, porque efectivamente dan resultados bien diferentes

Entiendo, es algo similar a la pregunta anterior. Es un buena idea usar otros canales de color, pero se considera para efectos la tarea la misma feature. Me gustaria que en la tarea prueben distintas familias de features.

**Comment by IchottMano** [comment]
Hola Profesor,

Quedé con la duda de cual es la tercera métrica que efectivamente nos piden, ya que en el enunciado dice "Tasa de falsos negativos (FPR)", supongo que es un error.

Saludos!

**Comment by domingomery** [comment]
> ... el enunciado dice "Tasa de falsos negativos (FPR)", supongo que es un error.

Efectivamente es un error, debe decir "Tasa de Falsos Positivos (FPR)", lo voy a corregir. Gracias por avisar!

**Comment by matipina** [comment]
Hola profe! Una pregunta, es válido realizar algún tipo de segmentación (aplicando Otsu por ejemplo) de las imágenes para dejarlas como imágenes binarias, y luego extraer características de estas?

Saludos!

**Comment by domingomery** [comment]
> es válido realizar algún tipo de segmentación (aplicando Otsu por ejemplo) de las imágenes para dejarlas como imágenes binarias, y luego extraer características de estas?

Sí es válido. Puedes sacar características geométricas de la imagen binaria, y también puedes extraer características de textura en la imagen a color sólo en pixeles segementados.
